packs/tfe_fdo_nomad/README.md

```diff
@@ -0,0 +1,162 @@
+1. `NOMAD_CLIENT_CERT` - Path to a PEM encoded client certificate for TLS authentication to the Nomad server. Must also specify NOMAD_CLIENT_KEY.
+1. `NOMAD_CLIENT_KEY` - Path to an unencrypted PEM encoded private key matching the client certificate from NOMAD_CLIENT_CERT.
+
+After setting up the environment variables, the pack can be setup using the following steps:
+
+1. Create Namespace for TFE job and TFE agent job.
+
+   1. Run `nomad namespace apply terraform-enterprise` to create the `terraform-enterprise` namespace. This is the default namespace that is used to bring up TFE Job.
+   1. Run `nomad namespace apply tfe-agents` to create the `tfe-agents` namespace. This is the default namespace that is used to bring up TFE Agent Job.
+
+
+2. Create a Nomad ACL policy file `terraform_enterprise_policy.hcl` with the content below:
+```hcl
+  namespace "tfe-agents" {
+  capabilities = ["submit-job","dispatch-job", "list-jobs", "read-job", "read-logs" ]
+  }
+  ```
+
+3. Apply the Nomad policy using:
+  ```bash
+  $ nomad acl policy apply \
+   -namespace terraform-enterprise -job tfe-job \
+   -group tfe-group -task tfe-task \
+   terraform-enterprise-policy ./terraform_enterprise_policy.hcl
+  ``` 
+
+4. Create the necessary Nomad Variables for each job. 
+  
+  These contain sensitive data that are required like certs, licenses and passwords.
+  Create a variable specification file: 
+
+  ```hcl
+  # spec.nv.hcl
+
+  # Path where Nomad variables will be stored, the same path will be used inside TFE job file rendered by tfe.nomad.tpl for TFE job to access.
+  path      = "nomad/jobs/tfe-job"
+  namespace = "terraform-enterprise"
+
+  items {
+    # TFE DB password. Mapped to the TFE_DB_PASSWORD environment variable.
+    db_password = ""
+
+    # The field should contain the base64 encoded value of the cert. Mappped to the TFE_TLS_CERT_FILE environment variable.
+    cert = ""
+
+    # The field should contain the base64 encoded value of the bundle. Mapped to the TFE_TLS_CA_BUNDLE_FILE environment variable.
+    bundle = ""
+
+    # The field should contain the base64 encoded value of the key. Mappped to the TFE_TLS_KEY_FILE environment variable.
+    key = ""
+  
+    # A valid TFE license. Mapped to the TFE_LICENSE environment variable.
+    tfe_license = ""
+    
+    # Object storage access key. Mapped to the TFE_OBJECT_STORAGE_S3_SECRET_ACCESS_KEY environment variable.
+    s3_secret_key = ""
+
+    # The field should contain the base64 encoded value of the Nomad CA. Mapped to the TFE_RUN_PIPELINE_NOMAD_TLS_CONFIG_CA_CERT environment variable.
+    nomad_ca_cert = ""
+
+    # The field should contain the base64 encoded value of the Nomad cert. Mapped to the TFE_RUN_PIPELINE_NOMAD_TLS_CONFIG_CLIENT_CERT environment variable.
+    nomad_cert = ""
+
+    # The field should contain the base64 encoded value of the Nomad cert's key. Mapped to the TFE_RUN_PIPELINE_NOMAD_TLS_CONFIG_CLIENT_KEY environment variable.
+    nomad_cert_key = ""
```

Comment:
This is so that TFE agent can talk to Nomad, right? 

There's no `NOMAD_TOKEN` being set anywhere, which makes me think we should be using the [Task API socket](https://developer.hashicorp.com/nomad/api-docs/task-api) and not TLS certificates for having TFE talk to Nomad anyways. But the Pack doesn't include what Nomad ACL policies need to be written for the TFE job either.

Reply:
> This is so that TFE agent can talk to Nomad, right?
TFE Agent is not supposed to talk to Nomad. Only the TFE container will.

> There's no `NOMAD_TOKEN` being set anywhere, which makes me think we should be using the [Task API socket](https://developer.hashicorp.com/nomad/api-docs/task-api) and not TLS certificates for having TFE talk to Nomad anyways. But the Pack doesn't include what Nomad ACL policies need to be written for the TFE job either.
`NOMAD_TOKEN` will be injected through `identity` stanza in TFE job file (not the TFE agent job file) if I'm not wrong. For versions not supporting `identity` we can put `NOMAD_TOKEN` in the job file.

Reply:
Ok, good to hear that you're using Workload Identity, but then you don't need the certs to use the Task API socket.

> If you meant to update the job definition to use unix.sock we can do a spike on it before GA and update. Note, that this is beta.

I was told this was urgently shipping to production and that's why we needed it to be at the top of our priority list to review. Let's just finish the job.

Reply:
@tgross if I am not wrong, Task APIs are not a special type of job, right? It's an HTTP API which we can call to start a new job on Nomad.

I'm not sure how we can update the job specs here to accommodate Task API.

The agent job here is started by TFE task worker (https://github.com/hashicorp/tfe-task-worker/blob/main/driver/nomad/nomad.go).

Reply:
The Task API provides a Unix domain socket at `${NOMAD_SECRETS_DIR}/api.sock` which you'd use instead of the external-facing HTTP endpoint. So no TLS required, just the Workload Identity token.

Reply:
yes, correct, so essentially we will have to modify the backend HTTP call to trigger TFE agent and not the job file.

Reply:
Why would you need to change the backend? You just need to swap out the `NOMAD_ADDR` environment variable. See https://github.com/hashicorp/nomad/pull/16872 for an example.

Reply:
oh ok, now it makes sense! thanks @tgross

Reply:
@tgross this won't work with TFE because of this code - https://github.com/hashicorp/terraform-enterprise/blob/b9f00dc90468660cd9ab99ff5334b9504d020974/config/config.go#L476

I'll have to pick this up, change, test and merge before the next release cycle.

Reply:
Very well. I'll approve.